docs/Tutorial_Contrast_Profile2.ipynb

```diff
@@ -0,0 +1,485 @@
+{
```

Comment:
<p>This section is the most important and I think we should move all of this to the top after you plot the ECG time series for the first time.</p><p> </p><p>@nimasarajpoor I wonder what would happen if we did a top-k nearest neighbor (with <code>k = 5</code>)and then plot the distances only for the 5th nearest neighbor (rather than the first 4 nearest neighbors). Perhaps, for periodic data with a single interesting pair of subsequences like this one, it would/could be a nice tutorial for understand top-k nearest neighbor matrix profiles. It is too much for this tutorial though.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/800/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Reply:
@seanlaw 
we can try top-k matrix profile and see its output for this data. I am having a hard time to picture the outcome in my mind in advance. So, what I am about to say is just based on some vague images in my mind. I have not thought this through. Please proceed with caution. 

We know that the top-k neighbors of a sequence, computed by `stumpy.stump`, are out of the trivial zone of the sequence, but the k neighbors might be in trivial neighbor of each other. So, for sequence `S_i`, we may find out that `S_j` is its 1NN, which is outside of the trivial zone of `i`, and say their distance is small (similar to the case provided in the notebook) It MIGHT BE possible that the 2nd, 3rd, ..., k-th NN of `S_i` are very close to `S_j`. For instanc, they might be `S_{j+1}` and `S_{j+2}`, .... So, the distance between `S_i` and its other top-k neighbors might be small as well. Therefore, if matrix profile `P` has `k` columns that correspond to top-k NN, `np.argmax(P[:, x])` may not necessarily give us the answer. Ofcourse, what I just said depends on the value of `k`. For a large enough (?) "k", we might find the desirable patterns (of the phenomena) via `np.argmax(P[:, x])`.  

We can try removing the trivial zone of 1NN, and find the next NN, out of that top-k NN. Then, `np.argmax(P')` might give us the answer, where `P'` is the distance to second nn AFTER IGNORING THE TRIVIAL ZONE OF 1NN of a subsequence (in additon to ignoring the trivial zone of the subsequence itself). The problem may become more challenging if we have more than two almost-similar-but-distinct-from-other patterns since now you have to choose a larger `k` if you want to consider the removal of trivial zone after computing the top-k neighbors. Note that for large window size `m`, the exclusion zone becomes bigger, and we may be forced to choose greater value for `k`.

To me, contrast profile can be seen as  a tool that allows user to put a filter on usual, regular patterns. So, for instance, say I have temperature data of a device and it is mostly as usual. But, there might be some strange behavior from time to time. How to find it? If it is a really unique, strange behavior, it may be discovered by discord, i.e. `np.argmax(P) `. But, what if it is not unique? One way might be to start from left and as you move to the right on the time axis, you only consider left matrix profile, and you detect an anomally as soon as you find one. Then, you can just remove it from your usual data, So, when you get to the next anomly, it can be detected (otherwise, it may not be detectable if it is similar/close to previous anomaly). So, what is the point of contrast profile? Maybe one use contrast profile to see if the new anomaly is something that appears before. In other words: repeated anomaly.

Again, it is better to do the experiment on the data and see what the data / result tell us.

Reply:
@ken-maeda To be clear, I am NOT suggesting that we do top-k here. It was simply an observation that this data set might be a good example for a separate top-k tutorial. Sorry for the distraction!